File: mobile/lib/models/product.dart
class ProductColor {
  final String name;
  final String hex;
  final String? swatchImage;
  final int stock;

  ProductColor({
    required this.name,
    required this.hex,
    this.swatchImage,
    this.stock = 0,
  });

  factory ProductColor.fromJson(Map<String, dynamic> json) => ProductColor(
        name: json['name'] ?? '',
        hex: json['hex'] ?? '#000000',
        swatchImage: json['swatchImage'],
        stock: json['stock'] ?? 0,
      );
}

class ProductFrame {
  final String? type;
  final String? material;
  final double? width;
  final double? lensWidth;
  final double? bridgeWidth;
  final double? templeLength;
  final List<String> featureTags;

  ProductFrame({
    this.type,
    this.material,
    this.width,
    this.lensWidth,
    this.bridgeWidth,
    this.templeLength,
    this.featureTags = const [],
  });

  factory ProductFrame.fromJson(Map<String, dynamic> json) => ProductFrame(
        type: json['type'],
        material: json['material'],
        width: (json['width'] as num?)?.toDouble(),
        lensWidth: (json['lensWidth'] as num?)?.toDouble(),
        bridgeWidth: (json['bridgeWidth'] as num?)?.toDouble(),
        templeLength: (json['templeLength'] as num?)?.toDouble(),
        featureTags: List<String>.from(json['featureTags'] ?? []),
      );
}

class ProductCompatible {
  final bool prescription;
  final bool bluecut;
  final bool zeropower;
  final bool progressive;

  ProductCompatible({
    this.prescription = false,
    this.bluecut = false,
    this.zeropower = false,
    this.progressive = false,
  });

  factory ProductCompatible.fromJson(Map<String, dynamic> json) => ProductCompatible(
        prescription: json['prescription'] ?? false,
        bluecut: json['bluecut'] ?? false,
        zeropower: json['zeropower'] ?? false,
        progressive: json['progressive'] ?? false,
      );
}

class Product {
  final String id;
  final String sku;
  final String name;
  final ProductFrame? frame;
  final List<ProductColor> colors;
  final List<String> images;
  final double originalPrice;
  final double sellingPrice;
  final double rating;
  final int reviewCount;
  final int soldCount;
  final List<String> categories;
  final bool isBestseller;
  final bool isActive;
  final ProductCompatible? compatible;

  Product({
    required this.id,
    required this.sku,
    required this.name,
    this.frame,
    this.colors = const [],
    this.images = const [],
    this.originalPrice = 999,
    this.sellingPrice = 1,
    this.rating = 0,
    this.reviewCount = 0,
    this.soldCount = 0,
    this.categories = const [],
    this.isBestseller = false,
    this.isActive = true,
    this.compatible,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    final price = json['price'] as Map<String, dynamic>?;
    return Product(
      id: json['_id'] ?? json['id'] ?? '',
      sku: json['sku'] ?? '',
      name: json['name'] ?? '',
      frame: json['frame'] != null ? ProductFrame.fromJson(json['frame']) : null,
      colors: (json['colors'] as List<dynamic>?)
              ?.map((c) => ProductColor.fromJson(c))
              .toList() ??
          [],
      images: List<String>.from(json['images'] ?? []),
      originalPrice: (price?['original'] as num?)?.toDouble() ?? 999,
      sellingPrice: (price?['selling'] as num?)?.toDouble() ?? 1,
      rating: (json['rating'] as num?)?.toDouble() ?? 0,
      reviewCount: json['reviewCount'] ?? 0,
      soldCount: json['soldCount'] ?? 0,
      categories: List<String>.from(json['categories'] ?? []),
      isBestseller: json['isBestseller'] ?? false,
      isActive: json['isActive'] ?? true,
      compatible: json['compatible'] != null
          ? ProductCompatible.fromJson(json['compatible'])
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        '_id': id,
        'sku': sku,
        'name': name,
        'price': {'original': originalPrice, 'selling': sellingPrice},
        'rating': rating,
        'reviewCount': reviewCount,
        'isBestseller': isBestseller,
      };
}

File: server/src/models/Product.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IProductColor {
  name: string;
  hex: string;
  swatchImage?: string;
  images: string[];
  stock: number;
}

export interface IProduct extends Document {
  sku: string;
  name: string;
  description?: string;
  frame: {
    type: string;
    material: string;
    width?: number;
    lensWidth?: number;
    bridgeWidth?: number;
    templeLength?: number;
    featureTags: string[];
  };
  frameType?: string;
  material?: string;
  brand?: string;
  shape?: string;
  frameSize?: string;
  frameColor?: string;
  weight?: string;
  faceShapes?: string[];
  isPremium?: boolean;
  tryIn3D?: boolean;
  colors: IProductColor[];
  defaultColor?: string;
  images: string[];
  image360?: string;
  price: {
    original: number;
    selling: number;
  };
  mrp?: number;
  sellingPrice?: number;
  discountPercent?: number;
  category: string;
  categories: string[];
  gender?: 'men' | 'women' | 'kids' | 'unisex';
  compatible: {
    prescription: boolean;
    bluecut: boolean;
    zeropower: boolean;
    progressive: boolean;
  };
  lensCompatibility: string[];
  tags: string[];
  rating: number;
  reviewCount: number;
  soldCount: number;
  weeklyBought?: number;
  isBestseller: boolean;
  isActive: boolean;
  meta: {
    seoTitle?: string;
    seoDescription?: string;
  };
  createdAt: Date;
  updatedAt: Date;
}

const ProductColorSchema = new Schema<IProductColor>({
  name: String,
  hex: String,
  swatchImage: String,
  images: [String],
  stock: { type: Number, default: 0 },
});

const ProductSchema = new Schema<IProduct>(
  {
    sku: { type: String, unique: true, required: true },
    name: { type: String, required: true },
    description: { type: String },
    frame: {
      type: {
        type: String,
        description: 'Frame type like Square, Round, etc.',
      },
      material: String,
      width: Number,
      lensWidth: Number,
      bridgeWidth: Number,
      templeLength: Number,
      featureTags: [String],
    },
    frameType: { type: String },
    material: { type: String },
    brand: { type: String },
    shape: { type: String },
    frameSize: { type: String },
    frameColor: { type: String },
    weight: { type: String },
    faceShapes: [{ type: String }],
    isPremium: { type: Boolean, default: false },
    tryIn3D: { type: Boolean, default: false },
    colors: [ProductColorSchema],
    defaultColor: { type: String },
    images: [String],
    image360: { type: String },
    price: {
      original: { type: Number, default: 999 },
      selling: { type: Number, default: 1 },
    },
    mrp: { type: Number },
    sellingPrice: { type: Number },
    discountPercent: { type: Number },
    category: {
      type: String,
      enum: ['prescription', 'sunglasses', 'blue_light', 'contact_lenses', 'kids'],
    },
    categories: [String],
    gender: {
      type: String,
      enum: ['men', 'women', 'kids', 'unisex'],
      default: 'unisex',
    },
    compatible: {
      prescription: { type: Boolean, default: false },
      bluecut: { type: Boolean, default: false },
      zeropower: { type: Boolean, default: false },
      progressive: { type: Boolean, default: false },
    },
    lensCompatibility: [String],
    tags: [String],
    rating: { type: Number, default: 0, min: 0, max: 5 },
    reviewCount: { type: Number, default: 0 },
    soldCount: { type: Number, default: 0 },
    weeklyBought: { type: Number, default: 0 },
    isBestseller: { type: Boolean, default: false },
    isActive: { type: Boolean, default: true },
    meta: {
      seoTitle: String,
      seoDescription: String,
    },
  },
  { timestamps: true }
);

ProductSchema.index({ sku: 1 });
ProductSchema.index({ category: 1 });
ProductSchema.index({ categories: 1 });
ProductSchema.index({ name: 'text', tags: 'text' });
ProductSchema.index({ brand: 1 });
ProductSchema.index({ shape: 1 });
ProductSchema.index({ frameSize: 1 });
ProductSchema.index({ frameColor: 1 });
ProductSchema.index({ frameType: 1 });
ProductSchema.index({ material: 1 });
ProductSchema.index({ weight: 1 });
ProductSchema.index({ faceShapes: 1 });
ProductSchema.index({ rating: 1 });
ProductSchema.index({ isPremium: 1 });
ProductSchema.index({ tryIn3D: 1 });
ProductSchema.index({ gender: 1 });

export const Product = mongoose.models.Product || mongoose.model<IProduct>('Product', ProductSchema);
